src/coreclr/jit/valuenum.h

```diff
@@ -463,7 +471,7 @@ class ValueNumStore
 
 #ifdef FEATURE_SIMD
     // A helper function for constructing VNF_SimdType VNs.
-    ValueNum VNForSimdType(unsigned simdSize, var_types simdBaseType);
+    ValueNum VNForSimdType(unsigned simdSize, CorInfoType simdBaseJitType);
```

Comment:
Encoding the dissolved (`var_types`) type and rehydrating that back into `CorInfoType` seems the simplest solution to me.

Note that there is no way for VN to know today that `n[u]int` is equivalent to `[u]int/[u]long` because the tables don't encode "precise" base types.

Reply:
I don't understand why it's better to encode `var_types` rather than `CorInfoType` when it encodes it as a SIMD type. Today, nothing is decoding this information other than the dump.

Reply:
If that's the case, then this behavior was happening before (notice GetSimdBaseType):
(original encoding)
```cpp
ValueNum simdTypeVN = vnStore->VNForSimdType(tree->GetSimdSize(), tree->GetSimdBaseType());
```
```cpp
var_types GenTreeJitIntrinsic::GetSimdBaseType() const
{
    CorInfoType simdBaseJitType = GetSimdBaseJitType();

    if (simdBaseJitType == CORINFO_TYPE_UNDEF)
    {
        return TYP_UNKNOWN;
    }
    return JitType2PreciseVarType(simdBaseJitType);
}
```

Reply:
I now understand what you are all saying.

Basically, you want to keep `Vector128<long>` and `Vector<nint>` on x64 the same as they are equivalent, and you want value numbering to take advantage of that equivalence.

I'm going to write a test to verify this behavior and ensure we do not regress.

Reply:
I did an example:
```csharp
    [MethodImpl(MethodImplOptions.NoInlining)]
    static unsafe Vector128<long> AdvSimd_Arm64_CompareEqual_Vector128_Long_AsVariableLoop(Vector128<long> left)
    {
        Vector128<long> result = default;
        Vector128<long> asVar = Vector128.Create((long)0);
        Vector128<nint> asVar2 = Vector128.Create((nint)0);
        Vector128<long> asVar3 = asVar2.AsInt64();
        for (var i = 0; i < 4; i++)
        {
            result = AdvSimd.Arm64.CompareEqual(left, asVar);
            for (var j = 0; j < 4; j++)
            {
                result = AdvSimd.Arm64.CompareEqual(left, asVar3);
            }
        }
        return result;
    }
```

Here is the diff, and this is with encoding `CorInfoType`:
```diff
             mov     fp, sp
 						;; bbWeight=1    PerfScore 1.50
 G_M7175_IG02:        ; gcrefRegs=0000 {}, byrefRegs=0000 {}, byref
-            movi    v16.2d, #0x00
-            movi    v17.2d, #0x00
             mov     w0, wzr
-						;; bbWeight=1    PerfScore 1.50
+						;; bbWeight=1    PerfScore 0.50
 G_M7175_IG03:        ; gcrefRegs=0000 {}, byrefRegs=0000 {}, byref, isz, align
-            cmeq    v18.2d, v0.2d, v17.2d
+            cmeq    v16.2d, v0.2d, #0
             mov     w1, wzr
             align   [4 bytes for IG04]
+            align   [4 bytes]
+            align   [4 bytes]
             align   [0 bytes]
-            align   [0 bytes]
-            align   [0 bytes]
-						;; bbWeight=4    PerfScore 8.00
+						;; bbWeight=4    PerfScore 12.00
 G_M7175_IG04:        ; gcrefRegs=0000 {}, byrefRegs=0000 {}, loop=IG04, byref, isz
-            cmeq    v18.2d, v0.2d, v16.2d
+            cmeq    v16.2d, v0.2d, #0
             add     w1, w1, #1
             cmp     w1, #4
             blt     G_M7175_IG04
@@ -45,14 +43,14 @@ G_M7175_IG05:        ; gcrefRegs=0000 {}, byrefRegs=0000 {}, byref, isz
             blt     G_M7175_IG03
 						;; bbWeight=4    PerfScore 8.00
 G_M7175_IG06:        ; gcrefRegs=0000 {}, byrefRegs=0000 {}, byref
-            mov     v0.16b, v18.16b
+            mov     v0.16b, v16.16b
 						;; bbWeight=1    PerfScore 0.50
 G_M7175_IG07:        ; , epilog, nogc, extend
             ldp     fp, lr, [sp],#16
             ret     lr
 						;; bbWeight=1    PerfScore 2.00
```
Both `movi` instructions have disappeared and the optimization is working as intended. I'm wondering if `CorInfoType` is being normalized somewhere.

Reply:
Did another (I'm doing Create(1) instead of Create(0)):
```csharp
    [MethodImpl(MethodImplOptions.NoInlining)]
    static unsafe Vector128<long> AdvSimd_Arm64_CompareEqual_Vector128_Long_AsVariableLoop(Vector128<long> left)
    {
        Vector128<long> result = default;
        Vector128<long> asVar = Vector128.Create((long)1);
        Vector128<nint> asVar2 = Vector128.Create((nint)1);
        Vector128<long> asVar3 = asVar2.AsInt64();
        for (var i = 0; i < 4; i++)
        {
            result = AdvSimd.Arm64.CompareEqual(left, asVar);
            for (var j = 0; j < 4; j++)
            {
                result = AdvSimd.Arm64.CompareEqual(left, asVar3);
            }
        }
        return result;
    }
```
And I got no diffs for it.

Reply:
FWIW, here's a snippet that shows the encoding deficiency:
```cs
private static Vector128<long> Problem(Vector128<long>* v)
{
    return (*v + Vector128<nint>.Zero.AsInt64()) + (*v + Vector128<long>.Zero);
}
```
```asm
; With "var_types" base types
            ldr     q16, [x0]
            movi    v17.4s, #0x00
            add     v16.2d, v16.2d, v17.2d
            add     v16.2d, v16.2d, v16.2d
            mov     v0.16b, v16.16b

; With "CorInfoType" base types
            ldr     q16, [x0]
            movi    v17.4s, #0x00
            add     v17.2d, v16.2d, v17.2d
            movi    v18.4s, #0x00
            add     v16.2d, v16.2d, v18.2d
            add     v16.2d, v17.2d, v16.2d
            mov     v0.16b, v16.16b
```
(We lose a CSE)

I agree it seems unlikely to show up in user code. I just think it is confusing the the developers reading the Jit code - why would we want to deliberately encode more than what is required?

I see the latest set of changes have addresses this.

Reply:
I made a change here: https://github.com/dotnet/runtime/pull/65028/commits/b1c44cff65532780bc8199f8e2a9789f06e137e0

That will try to resolve this.

Reply:
Thank you for the feedback by the way; it's been very helpful. :)